src/celeritas/em/msc/UrbanMsc.hh

```diff
@@ -101,18 +101,40 @@ UrbanMsc::limit_step(CoreTrackView const& track, StepLimit* step_limit)
 {
     CELER_EXPECT(msc_params_);
 
-    auto geo = track.make_geo_view();
     auto phys = track.make_physics_view();
+    auto par = track.make_particle_view();
+    detail::UrbanMscHelper msc_helper(msc_params_, par, phys);
 
-    // TODO: limited is set as a side effect of the lambda below: change to be
-    // part of the MSC return result
-    bool limited = false;
+    bool displaced = false;
 
     // Sample multiple scattering step length
-    auto msc_step = [&] {
-        auto par = track.make_particle_view();
-        real_type const safety = (geo.is_on_boundary() ? 0 : geo.find_safety());
-        detail::UrbanMscHelper msc_helper(msc_params_, par, phys);
+    real_type const true_path = [&] {
+        if (step_limit->step <= msc_params_.params.limit_min_fix())
+        {
+            // Very short step: don't displace or limit
+            return step_limit->step;
+        }
+
+        auto geo = track.make_geo_view();
+
+        real_type safety = 0;
+        if (!geo.is_on_boundary())
+        {
+            // Because the MSC behavior changes based on whether the *total*
+            // track range is close to a boundary, rather than whether the next
+            // steps are closer, we need to find the safety distance up to the
+            // potential travel radius of the particle at its current energy.
+            real_type const max_step = msc_helper.max_step();
+            safety = geo.find_safety(max_step);
+            if (safety >= max_step)
+            {
+                // The nearest boundary is further than the maximum expected
+                // travel distance of the particle: don't displace or limit
```

Comment:
Yeah, agreed. I suppose the idea is that multiple scattering in an infinite medium will result in no average change in displacement, so it can be neglected far from boundaries? I imagine this could definitely have an effect in a sensitive detector that factors in the track's location (e.g. a region that represents an array of detector cells).